src/hidden.rs

```diff
@@ -20,74 +20,248 @@
+/// 
+/// // Clone the data safely
+/// let mut example_clone = example.clone();
+/// 
+/// // Zeroize the data manually
+/// example_clone.zeroize();
+/// assert_eq!(example_clone.reveal(), &[0u8; 32]);
+/// # }
+/// ```
+#[macro_export]
+macro_rules! hidden_type {
+    ($name:ident, $type:ty) => {
+        /// A hidden type
+        #[derive(Clone, Debug, Zeroize)]
+        pub struct $name
+        {
+            data: Hidden<$type>,
+        }
+
+        impl $name {
+            /// Get an immutable reference to the data
+            #[allow(dead_code)]
+            pub fn reveal(&self) -> &$type {
+                self.data.reveal()
+            }
+
+            /// Get a mutable reference to the data
+            #[allow(dead_code)]
+            pub fn reveal_mut(&mut self) -> &mut $type {
+                self.data.reveal_mut()
+            }
+        }
+
+        impl From<$type> for $name {
+            /// Hide existing data
+            fn from(t: $type) -> Self {
+                Self { data: Hidden::hide(t) }
+            }
+        }
+    };
+}
+
+/// A generic type for data that needs to be kept hidden and zeroized when out of scope, and is accessible only by reference.
+///
+/// You can define a hidden type using any underlying data type that implements `Zeroize` and serialization.
+/// This is the case for most basic types that you probably care about.
+///
+/// Hidden data has useful properties:
+/// - The data is not subject to `Debug` or `Display` output, which are masked.
+/// - The data can only be accessed by (immutable or mutable) reference, and cannot be copied.
+/// - The data zeroizes when dropped, and can also be manually zeroized.
+/// - Cloning is safe.
+/// - Serialization and deserialization are supported transparently.
+/// 
+/// ```edition2018
+/// # use rand::rngs::OsRng;
+/// # use rand::RngCore;
+/// # use tari_utilities::hidden::Hidden;
+/// # use zeroize::Zeroize;
+///
+/// // In this example, we need to handle secret data of type `[u8; 32]`.
+///
+/// // We can create hidden data from existing data; in this case, it's the caller's responsibility to make sure the existing data is handled securely
+/// let hidden_from_data = Hidden::<[u8; 32]>::hide([1u8; 32]);
+///
+/// // We can access the hidden data as a reference, but not take ownership of it
+/// assert_eq!(hidden_from_data.reveal(), &[1u8; 32]);
+///
+/// // We can create default hidden data and then modify it as a mutable reference; this is common for functions that act on data in place
+/// let mut rng = OsRng;
+/// let mut hidden_in_place = Hidden::<[u8; 32]>::hide([0u8; 32]);
+/// rng.fill_bytes(hidden_in_place.reveal_mut());
+///
+/// // Cloning is safe to do
+/// let mut clone = hidden_in_place.clone();
+/// assert_eq!(hidden_in_place.reveal(), clone.reveal());
+/// 
+/// // You can manually zeroize the data if you need to
+/// clone.zeroize();
+/// assert_eq!(clone.reveal(), &[0u8; 32]);
+/// ```
+#[derive(Clone, Deserialize, Serialize)]
```

Comment:
We go through all this effort of making the data safe, but here open ourselves up to inadvertently leaking the data via serialisation.

We can keep deserialisation, since the value is already in the clear.

For serialisation, one option is to just require users to provide a `serialize_with(...)` on fields that are `Hidden`. That's a burden, but maintains the level of data security we've held until now.